### PR TITLE
Optionally verify S3 artifacts via detached signatures

### PR DESCRIPTION
**Pull request by jonathanwgoodwin**

### CODE REVIEW

File: SingularityS3Base/src/main/java/com/hubspot/singularity/s3/base/ArtifactManager.java [hunk 1]
```diff
@@ -87,6 +97,37 @@ private void downloadAndCheck(RemoteArtifact artifact, Path downloadTo) {
 
     checkFilesize(artifact, downloadTo);
     checkMd5(artifact, downloadTo);
+    checkGpgSignature(artifact, downloadTo);
+  }
+
+  private void checkGpgSignature(RemoteArtifact baseArtifact, Path downloadTo) {
+    if (baseArtifact.getGpgSignatureArtifact().isPresent() && baseArtifact.getGpgSignatureArtifact().get() instanceof RemoteArtifact) {
```
Comment (wsorenson) [inline SingularityS3Base/src/main/java/com/hubspot/singularity/s3/base/ArtifactManager.java]: Always better to reduce indentation. If you have an if block around an entire method, you should either split the inner contents into its own method or negate the IF and return (my preference.)

File: SingularityS3Base/src/main/java/com/hubspot/singularity/s3/base/ArtifactManager.java [hunk 2]
```diff
@@ -87,6 +97,37 @@ private void downloadAndCheck(RemoteArtifact artifact, Path downloadTo) {
...
+
+      Path signatureArtifactPath =  fetch((RemoteArtifact) baseArtifact.getGpgSignatureArtifact().get());
```
Comment (wsorenson) [inline SingularityS3Base/src/main/java/com/hubspot/singularity/s3/base/ArtifactManager.java]: Casting is pretty sucky. Why not just add this field to the top-level artifact class if we're going to check it here?

File: SingularityS3Base/src/main/java/com/hubspot/singularity/s3/base/ArtifactManager.java [hunk 3]
```diff
@@ -87,6 +97,37 @@ private void downloadAndCheck(RemoteArtifact artifact, Path downloadTo) {
 
     checkFilesize(artifact, downloadTo);
     checkMd5(artifact, downloadTo);
+    checkGpgSignature(artifact, downloadTo);
+  }
+
+  private void checkGpgSignature(RemoteArtifact baseArtifact, Path downloadTo) {
```
Comment (wsorenson) [inline SingularityS3Base/src/main/java/com/hubspot/singularity/s3/base/ArtifactManager.java]: This method probably needs logging.

File: SingularityS3Base/src/main/java/com/hubspot/singularity/s3/base/config/SingularityS3Configuration.java [hunk 4]
```diff
@@ -67,6 +67,23 @@
...
+  @NotEmpty
+  @JsonProperty
+  private String gpgBinaryPath = "/usr/bin/gpg";
+
+  @NotEmpty
+  @Obfuscate
+  @JsonProperty
+  private String gpgKeyPassword;
+
+  @NotEmpty
```
Comment (wsorenson) [inline SingularityS3Base/src/main/java/com/hubspot/singularity/s3/base/config/SingularityS3Configuration.java]: As these are NotEmpty does it imply gpg is now a "required" part of the Singularity S3 configuration?

**Comment by ssalinas** [comment]
:ship:

**Pull request merged**